test/integration/dualstack/dualstack_test.go

```diff
@@ -0,0 +1,1336 @@
+			serviceType:    v1.ServiceTypeClusterIP,
+			clusterIPs:     []string{},
+			ipFamilies:     []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy: v1.IPFamilyPolicySingleStack,
+			expectError:    true,
+		},
+		{
+			name:           "Type ClusterIP - Server Allocated IP - IPv6 IPv4 Family - Policy Prefer Dual Stack",
+			serviceType:    v1.ServiceTypeClusterIP,
+			clusterIPs:     []string{},
+			ipFamilies:     []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy: v1.IPFamilyPolicyPreferDualStack,
+			expectError:    true,
+		},
+		{
+			name:           "Type ClusterIP - Server Allocated IP - IPv6 IPv4 Family - Policy Required Dual Stack",
+			serviceType:    v1.ServiceTypeClusterIP,
+			clusterIPs:     []string{},
+			ipFamilies:     []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy: v1.IPFamilyPolicyRequireDualStack,
+			expectError:    true,
+		},
+	}
+
+	for i, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+
+			svc := &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("svc-test-%d", i), // use different services for each test
+				},
+				Spec: v1.ServiceSpec{
+					Type:           tc.serviceType,
+					ClusterIPs:     tc.clusterIPs,
+					IPFamilies:     tc.ipFamilies,
+					IPFamilyPolicy: &tc.ipFamilyPolicy,
+					Ports: []v1.ServicePort{
+						{
+							Name:       fmt.Sprintf("port-test-%d", i),
+							Port:       443,
+							TargetPort: intstr.IntOrString{IntVal: 443},
+							Protocol:   "TCP",
+						},
+					},
+				},
+			}
+
+			// create the service
+			_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+			if (err != nil) != tc.expectError {
+				t.Errorf("Test failed expected result: %v received %v ", tc.expectError, err)
+			}
+			// if no error was expected validate the service otherwise return
+			if err != nil {
+				return
+			}
+			// validate the service was created correctly if it was not expected to fail
+			svc, err = client.CoreV1().Services(metav1.NamespaceDefault).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("Unexpected error to get the service %s %v", svc.Name, err)
+			}
+			// if no IPFamily specified it must use the default (first cidr in service subnet)
+			ipFamilies := tc.ipFamilies
+			if len(ipFamilies) == 0 {
+				ipFamilies = []v1.IPFamily{defaultIPFamily}
+			}
+			if err := validateServiceAndClusterIPFamily(svc, ipFamilies); err != nil {
+				t.Errorf("Unexpected error validating the service %s\n%+v\n%v", svc.Name, svc, err)
+			}
+		})
+	}
+}
+
+// TestCreateServiceDualStackIPv6 test the Service dualstackness in an IPv6 only DualStack cluster
+func TestCreateServiceDualStackIPv6(t *testing.T) {
+	// Create an IPv6 only dual stack control-plane
+	serviceCIDR := "2001:db8:1::/48"
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, true)()
```

Comment:
minor but all this setup is boilerplate.  Why not  `client, err := setup(func(cfg *IntegrationConfig) { cfg.ExtraConfig.ServiceIPRange = *cidr })`  or even  `client, err := setup(setServiceIPRange(cidr))` - it's a pattern that makes for even easier to read tests (IMO).  See https://github.com/kubernetes/kubernetes/pull/95196/files#diff-f8cc365ab0a790d6235699f1f9bf8334c70b1be1ee970f8c51d5eef1b640ab53

Reply:
>  it's a pattern that makes for even easier to read tests (IMO)

I agree with you, but it depends on the the audience, right? 
I've found that this way is easier to understand for different people profiles and allows to copy-paste the code for creating new tests easily.